src/main/java/co/ivi/jus/record/review/xuling/SocialSecurityNumber.java

```
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) 2021, Ling Xu. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ */
+
+package co.ivi.jus.record.review.xuling;
+
+import java.util.Arrays;
+
+/**
+ * record class work
+ *
+ * @author jack
+ * @since 2021/11/24
+ **/
+public record SocialSecurityNumber(byte[] ssn) {
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+
+        if (o == null) {
+            return false;
+        }
+
+        if (o instanceof SocialSecurityNumber that) {
+            return Arrays.equals(ssn, that.ssn());
+        }
```

这里的实现是这个思考题最难的地方。敏感信息的比较，应该是时间恒定的。也就是说，信息比较计算的时间，只和信息的长度有关，而不能和信息的内容有关系。所以，建议使用java.security.MessageDigest. isEqual(byte[] digesta,           byte[] digestb)来比较敏感信息。

---

好的，这个主要是为了防止黑客攻击。当时感觉这个没有这么重要，按照这个逻辑，toString方法是不是也需要加密输出。

---

toString有待商榷，要看使用场景怎么用toString了。安全起见，也不建议输出SSN。倒是没有必要加密。

---

> 涨姿势
Arrays.equals() is fail fast method.
If there is a hash or key comparison using Arrays.equals(), the attack can fake a hash or key and do the comparison with the actual hash or key. They can measure the time taken for Arrays.equals() to return to understand how many bytes matched the faked hash or key has. After hundreds or thousands of tries, they may reveal the actual hash and key.

---

是否可以脱敏显示呢？

---

脱敏显示当然可以。